src/python/foglamp/device/ingest.py

```diff
@@ -0,0 +1,497 @@
+    async def add_readings(cls, asset: str, timestamp: Union[str, datetime.datetime],
+                           key: Union[str, uuid.UUID] = None, readings: dict = None)->None:
+        """Adds an asset readings record to FogLAMP
+
+        Args:
+            asset: Identifies the asset to which the readings belong
+            timestamp: When the readings were taken
+            key:
+                Unique key for these readings. If this method is called multiple with the same
+                key, the readings are only written to the database once
+            readings: A dictionary of sensor readings
+
+        Raises:
+            If this method raises an Exception, the discarded readings counter is
+            also incremented.
+
+            RuntimeError:
+                The server is stopping or has been stopped
+
+            ValueError, TypeError:
+                An invalid value was provided
+        """
+        if cls._stop:
+            raise RuntimeError('The device server is stopping')
+        # Assume the code beyond this point doesn't 'await'
+        # to make sure that the queue is not appended to
+        # when cls._stop is True
+
+        if not cls._started:
+            raise RuntimeError('The device server was not started')
+            # cls._logger = logger.setup(__name__, destination=logger.CONSOLE, level=logging.DEBUG)
+
+        try:
+            if asset is None:
+                raise ValueError('asset can not be None')
+
+            if not isinstance(asset, str):
+                raise TypeError('asset must be a string')
+
+            if timestamp is None:
+                raise ValueError('timestamp can not be None')
+
+            if not isinstance(timestamp, datetime.datetime):
+                # validate
+                timestamp = dateutil.parser.parse(timestamp)
+
+            if key is not None and not isinstance(key, uuid.UUID):
+                # Validate
+                if not isinstance(key, str):
+                    raise TypeError('key must be a uuid.UUID or a string')
+                # If key is not a string, uuid.UUID throws an Exception that appears to
+                # be a TypeError but can not be caught as a TypeError
+                key = uuid.UUID(key)
+
+            if readings is None:
+                readings = dict()
+            elif not isinstance(readings, dict):
+                # Postgres allows values like 5 be converted to JSON
+                # Downstream processors can not handle this
+                raise TypeError('readings must be a dictionary')
+        except Exception:
+            cls.increment_discarded_readings()
+            raise
+
+        # Comment out to test IntegrityError
+        # key = '123e4567-e89b-12d3-a456-426655440000'
+
+        cls.is_available()  # Locate a queue that isn't maxed out
+        queue_index = cls._current_readings_queue_index
+        queue = cls._readings_queues[queue_index]
+
+        if not cls._queue_readings_as_dict:
```

Comment:
if not ? and is this check actually required as we are checking and using json.dumps when fetching item for queue as an object to insert

Reply:
This point of this variable is to control whether json.dumps is called in add_readings or in _insert_readings. Benchmarks show that it should be done in _insert_readings. add_readings should be as fast as possible.